### ISSUE TITLE
issue with loading optifabric

### ISSUE THREAD

**Issue by BugIssuesAmoungUs** [Chocohead/OptiFabric#922]
i tried loading 1.19.2, and it just crashes, but doesnt leave a crash log, and i dont know the issue
[latest.log](https://github.com/Chocohead/OptiFabric/files/10095676/latest.log)

**Comment by topia** [comment]
```
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector removeComputeFloat(Lit/unimi/dsi/fastutil/objects/Object2FloatMap;Ljava/lang/Object;Ljava/util/function/BiFunction;)F in soundcategories.mixins.json:GameOptionsMixin from mod soundcategories failed injection check, (0/1) succeeded. Scanned 1 target(s). Using refmap soundcategories-refmap.json
```
Your latest.log shows that it failed with the soundcategories mod. so it seems like the mod compatibility issue between optifine/optifabric and soundcategories.

**Comment by Kwisma** [comment]
Loading 1.19.2 crashes, I don't know what's the problem
[minecraft-exported-crash-info-2022-11-27T18-37-49.zip](https://github.com/Chocohead/OptiFabric/files/10098044/minecraft-exported-crash-info-2022-11-27T18-37-49.zip)
[minecraft-exported-logs-2022-11-27T18-43-35.log](https://github.com/Chocohead/OptiFabric/files/10098058/minecraft-exported-logs-2022-11-27T18-43-35.log)

**Comment by topia** [comment]
> Loading 1.19.2 crashes, I don't know what's the problem [minecraft-exported-crash-info-2022-11-27T18-37-49.zip](https://github.com/Chocohead/OptiFabric/files/10098044/minecraft-exported-crash-info-2022-11-27T18-37-49.zip)

Please read error log.
```
[18:37:41] [main/ERROR]: Incompatible mod set!
net.fabricmc.loader.impl.FormattedException: Mod resolution encountered an incompatible mod set!
A potential solution has been determined:
	 - Remove mod 'OptiFabric' (optifabric) 1.13.16 (D:\hmcl\.minecraft\versions\李芒果空岛\mods\optifabric-1.13.16.jar).
Unmet dependency listing:
	 - Mod 'Sodium' (sodium) 0.4.4+build.18 is incompatible with any version of mod 'OptiFabric' (optifabric), but a matching version is present: 1.13.16!
	 - Mod 'Iris' (iris) 1.4.2 is incompatible with any version of mod 'OptiFabric' (optifabric), but a matching version is present: 1.13.16!
	 - Mod 'Lithium' (lithium) 0.10.2 is incompatible with any version of mod 'OptiFabric' (optifabric), but a matching version is present: 1.13.16!
```
Optifine/Optifabric and Sodium/Iris/Lithium are not compatible by design. so please remove either.

**Comment by SageTheSnail** [comment]
> ```
> Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Redirector removeComputeFloat(Lit/unimi/dsi/fastutil/objects/Object2FloatMap;Ljava/lang/Object;Ljava/util/function/BiFunction;)F in soundcategories.mixins.json:GameOptionsMixin from mod soundcategories failed injection check, (0/1) succeeded. Scanned 1 target(s). Using refmap soundcategories-refmap.json
> ```
> 
> Your latest.log shows that it failed with the soundcategories mod. so it seems like the mod compatibility issue between optifine/optifabric and soundcategories.

I'm having a similar error, but the critical injection failure was in porting_lib_base.mixins.json:client.LevelRendererMixin from mod porting_lib_base. Is this actually from a "mod" or is Optifine interacting with my modloader?

**Comment by topia** [comment]
> I'm having a similar error, but the critical injection failure was in porting_lib_base.mixins.json:client.LevelRendererMixin from mod porting_lib_base. Is this actually from a "mod" or is Optifine interacting with my modloader?

That failure was caused when the fabric loader failed to apply a patch (mixin) to a class. the reason is varied, but basically a compatibility problem.

Basically, *I* don't say that this is the problem of a single side. this is a compatibility problem between two mods (optifine/optifabric and another).
Optifabric replaces vanilla Minecraft java classes with Optifine patched classes for its purpose.
But ordinally, other mods assume that they're patching to vanilla (or only fabric-patched) Minecraft class.

Optifine might change method signature, internal implementation, etc. So an optifine class and a vanilla class might be incompatible.
For tackling the problem, Optifabric has some patches for *other mods, including fabric API*; developers (Chocohead and others, not me) carefully developed them.

This is the reason that the fabric mod developer recommends Sodium/Iris/Lithium, not optifabric. optifabric resolves the big problem for users and is created carefully, but it's fragile and complicated.

**Comment by SageTheSnail** [comment]
> > I'm having a similar error, but the critical injection failure was in porting_lib_base.mixins.json:client.LevelRendererMixin from mod porting_lib_base. Is this actually from a "mod" or is Optifine interacting with my modloader?
> 
> That failure was caused when the fabric loader failed to apply a patch (mixin) to a class. the reason is varied, but basically a compatibility problem.
> 
> Basically, _I_ don't say that this is the problem of a single side. this is a compatibility problem between two mods (optifine/optifabric and another). Optifabric replaces vanilla Minecraft java classes with Optifine patched classes for its purpose. But ordinally, other mods assume that they're patching to vanilla (or only fabric-patched) Minecraft class.
> 
> Optifine might change method signature, internal implementation, etc. So an optifine class and a vanilla class might be incompatible. For tackling the problem, Optifabric has some patches for _other mods, including fabric API_; developers (Chocohead and others, not me) carefully developed them.
> 
> This is the reason that the fabric mod developer recommends Sodium/Iris/Lithium, not optifabric. optifabric resolves the big problem for users and is created carefully, but it's fragile and complicated.

yep, believe you commented on my post about the issue. it took a bunch of trial and error since i can only read tiny amounts of code (basically only the stuff that's caused me crashes and was then explained to me by somebody) but I figured out what the mod interaction was.

**Comment by topia** [comment]
> I figured out what the mod interaction was.

In addition, I explained it as optifabric and others, but it also applies to almost all fabric mods. Fabric mods, except no-mixin mods, patch Minecraft code in their way. Sometimes, they patch the same place incompatibility way, and it causes a mod compatibility problem.
